File: src/algorithms/sets/generation/k-combinations/k-combinations.test.ts
import { describe, it, expect } from "vitest";
import { kCombinations } from "./sources/k-combinations.ts?fn";

describe("kCombinations", () => {
  it("generates C(5,3) = 10 combinations for the default input", () => {
    const result = kCombinations([1, 2, 3, 4, 5], 3) as number[][];
    expect(result.length).toBe(10);
  });

  it("every result subset has exactly k elements", () => {
    const result = kCombinations([1, 2, 3, 4, 5], 3) as number[][];
    for (const subset of result) {
      expect(subset.length).toBe(3);
    }
  });

  it("returns all expected combinations for [1,2,3,4,5] choose 3", () => {
    const result = kCombinations([1, 2, 3, 4, 5], 3) as number[][];
    const serialized = result
      .map((subset) => [...subset].sort((numA, numB) => numA - numB).join(","))
      .sort();
    const expected = [
      "1,2,3",
      "1,2,4",
      "1,2,5",
      "1,3,4",
      "1,3,5",
      "1,4,5",
      "2,3,4",
      "2,3,5",
      "2,4,5",
      "3,4,5",
    ].sort();
    expect(serialized).toEqual(expected);
  });

  it("returns C(4,2) = 6 combinations", () => {
    const result = kCombinations([1, 2, 3, 4], 2) as number[][];
    expect(result.length).toBe(6);
  });

  it("returns 1 combination when k equals n (full set)", () => {
    const result = kCombinations([1, 2, 3], 3) as number[][];
    expect(result.length).toBe(1);
    const subset = [...result[0]!].sort((numA, numB) => numA - numB);
    expect(subset).toEqual([1, 2, 3]);
  });

  it("returns 1 combination when k equals 1 (each element alone)", () => {
    const result = kCombinations([5, 10, 15], 1) as number[][];
    expect(result.length).toBe(3);
    for (const subset of result) {
      expect(subset.length).toBe(1);
    }
  });

  it("returns empty array when k = 0", () => {
    // k = 0 means no subsets of size 0 beyond the empty one
    // behavior: backtrack(0) immediately hits base case with empty subset
    const result = kCombinations([1, 2, 3], 0) as number[][];
    expect(result.length).toBe(1);
    expect(result[0]).toEqual([]);
  });

  it("returns empty array when k exceeds n", () => {
    const result = kCombinations([1, 2], 5) as number[][];
    expect(result.length).toBe(0);
  });

  it("returns empty array for empty input with k > 0", () => {
    const result = kCombinations([], 2) as number[][];
    expect(result.length).toBe(0);
  });

  it("produces no duplicate combinations", () => {
    const result = kCombinations([1, 2, 3, 4, 5], 3) as number[][];
    const serialized = result.map((subset) => [...subset].sort().join(","));
    const unique = new Set(serialized);
    expect(unique.size).toBe(result.length);
  });

  it("each subset contains only elements from the input", () => {
    const input = [10, 20, 30, 40];
    const result = kCombinations(input, 2) as number[][];
    for (const subset of result) {
      for (const value of subset) {
        expect(input).toContain(value);
      }
    }
  });
});

File: src/algorithms/sets/generation/k-combinations/__tests__/KCombinations_test.cpp
#define TESTING
#include "../sources/KCombinations.cpp"
#include <cassert>
#include <algorithm>
#include <set>
#include <string>
#include <iostream>

int main() {
    // C(5,3) = 10
    auto result1 = kCombinations({1, 2, 3, 4, 5}, 3);
    assert(result1.size() == 10);

    // every subset has exactly k elements
    for (const auto& subset : result1) {
        assert(subset.size() == 3);
    }

    // C(4,2) = 6
    auto result2 = kCombinations({1, 2, 3, 4}, 2);
    assert(result2.size() == 6);

    // k equals n — full set, exactly 1 result
    auto result3 = kCombinations({1, 2, 3}, 3);
    assert(result3.size() == 1);

    // k = 0 returns one empty subset
    auto result4 = kCombinations({1, 2, 3}, 0);
    assert(result4.size() == 1);
    assert(result4[0].empty());

    // k exceeds n — no combinations
    auto result5 = kCombinations({1, 2}, 5);
    assert(result5.empty());

    // empty input with positive k
    auto result6 = kCombinations({}, 2);
    assert(result6.empty());

    // no duplicate combinations
    std::set<std::string> uniqueSubsets;
    for (auto subset : result1) {
        std::sort(subset.begin(), subset.end());
        std::string key;
        for (int val : subset) key += std::to_string(val) + ",";
        uniqueSubsets.insert(key);
    }
    assert(uniqueSubsets.size() == result1.size());

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
